### ISSUE TITLE
`typing_extensions.deprecated` and `warnings.deprecated` remove coroutine property; How to deprecate async functions?

### ISSUE THREAD

**Issue by Kakadus** [python/typing_extensions#435]
When I annotate an async function with `@deprecated(...)`, `inspect.iscoroutinefunction` no longer returns True. This is not what I expect to happen, as with `partial` from `functools` this does not happen.

Consider the following code:

```python
import inspect

from typing_extensions import deprecated

async def coroutinefunction_a():
    pass

@deprecated("deprecated coroutine")
async def coroutinefunction_b():
     pass

assert inspect.iscoroutinefunction(coroutinefunction_a) is True  # correct
assert inspect.iscoroutinefunction(partial(coroutinefunction_b)) is True # correct
assert inspect.iscoroutinefunction(coroutinefunction_b) is True # fail
```

With typing_extensions and python3.13, the assertion fails.

This is especially annoying since e.g. [libraries use `inspect.iscoroutinefunction` to detect the type of function](https://github.com/miguelgrinberg/python-socketio/blob/main/src/socketio/async_namespace.py#L35-L41).

I am not sure this is the correct place for this bug report, as this both happens with python 3.13 and typing_extensions on older python versions.

**Comment by srittau** [comment]
`inspect.iscoroutinefunction()` has special handling for `partial()`. The real culprit here is `@functools.wraps`, which `@deprecated` ultimately uses:

```python
import inspect

from functools import wraps


async def coroutinefunction_a():
    pass

def wrapper(f):
    @wraps(f)
    def x():
        return f()

    return x

@wrapper
async def coroutinefunction_b():
    pass


assert inspect.iscoroutinefunction(coroutinefunction_a) is True  # correct
assert inspect.iscoroutinefunction(coroutinefunction_b) is True  # fail
```

`@wraps` (or rather `update_wrapper`) should probably copy the coroutine marker (`_is_coroutine_marker`) that `iscoroutinefunction` checks for.

**Comment by AlexWaygood** [comment]
> `@wraps` (or rather `update_wrapper`) should probably copy the coroutine marker (`_is_coroutine_marker`) that `iscoroutinefunction` checks for.

Previously discussed in:
- https://github.com/python/cpython/issues/100317

**Comment by srittau** [comment]
In that case, I definitely see the point of copying the flag in `@deprecated`. (Maybe `update_wrapper`/`@wraps` should get a flag to copy it, but that's off-topic for this issue?)

**Comment by AlexWaygood** [comment]
> In that case, I definitely see the point of copying the flag in `@deprecated`. (Maybe `update_wrapper`/`@wraps` should get a flag to copy it, but that's off-topic for this issue?)

Yeah, that sounds good to me. Even if the flag is added to `update_wrapper`/`@wraps`, it won't be added to the stdlib before Python 3.14, so we'll need an intermediate solution here.

**Comment by srittau** [comment]
@AlexWaygood @JelleZijlstra Could one of you move this issue over to the cpython repository? This should probably be fixed in CPython first.
